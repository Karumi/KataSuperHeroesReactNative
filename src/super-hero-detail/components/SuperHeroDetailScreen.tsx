import { fromNullable, Option } from "fp-ts/lib/Option";
import * as React from "react";
import { ScrollView, StyleSheet, Text, View } from "react-native";
import { NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch, Action } from "redux";
import EmptyCase from "../../base-components/empty-case/EmptyCase";
import Loading from "../../base-components/loading/Loading";
import navigationOptions from "../../base-components/navigationOptions";
import { SuperHero } from "../../core/model";
import { RootState } from "../../store";
import SuperHeroCell from "../../super-heroes-list/components/SuperHeroCell";
import { clearSuperHero, fetchSuperHeroById } from "../super-hero-detail-actions";

interface NavigationParams {
    title: string;
}

interface Props {
    readonly navigation: any;
    readonly onMount: (superHeroId: string) => void;
    readonly onUnmount: () => void;
    readonly loading: boolean;
    readonly superHero: Option<SuperHero>;
}
class SuperHeroDetailScreen extends React.Component<Props> {

    public static navigationOptions = ({ navigation }: NavigationScreenProps<NavigationParams>) =>
        navigationOptions(fromNullable(navigation.state.params).map((p: NavigationParams) => p.title).getOrElse(""))

    public componentWillMount() {
        const superHeroId = this.props.navigation.getParam("superHeroId");
        this.props.onMount(superHeroId);
    }

    public componentWillUnmount() {
        this.props.onUnmount();
    }

    public componentWillReceiveProps(nextProps: Props) {
        if (this.props.superHero.isNone() && nextProps.superHero.isSome()) {
            const superHero = nextProps.superHero;
            const navigationBarTitle = superHero.map((sh) => sh.name).getOrElse("");
            setTimeout(() => {
                nextProps.navigation.setParams({ title: navigationBarTitle });
            });
        }
    }

    public render() {
        const { loading, superHero } = this.props;
        const doesTheSuperHeroExist = superHero.isSome();
        const shouldShowEmptyCase = !loading && !doesTheSuperHeroExist;
        return (
            <View
                style={styles.screen}>
                {loading && <Loading />}
                {shouldShowEmptyCase && <EmptyCase />}
                {superHero.fold(null, (sh) => this.renderSuperHero(sh))}
            </View>
        );
    }

    private renderSuperHero(superHero: SuperHero) {
        return (
            <ScrollView>
                <SuperHeroCell onTap={() => { }} superHero={superHero} />
                <Text style={styles.description}>{superHero.description}</Text>
            </ScrollView>
        );
    }
}

const styles = StyleSheet.create({
    screen: {
        flex: 1,
        backgroundColor: "#22282F",
    },
    description: {
        color: "white",
        margin: 16,
        fontSize: 16,
    },
});

const mapStateToProps = (state: RootState) => ({
    loading: state.superHeroDetail.loading,
    superHero: state.superHeroDetail.superHero,
});

const mapDispatchToProps = (dispatch: Dispatch<Action>) => bindActionCreators({
    onMount: fetchSuperHeroById,
    onUnmount: clearSuperHero,
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(SuperHeroDetailScreen);
