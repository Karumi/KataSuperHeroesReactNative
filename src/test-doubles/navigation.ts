
export const mockPlainNavigation = {
    navigate: jest.fn(),
};

export const navigationWithMockedParams = {
    navigate: jest.fn(),
    getParam: jest.fn(),
};
