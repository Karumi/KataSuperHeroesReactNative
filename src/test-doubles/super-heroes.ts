import { SuperHero } from "../core/model";

export const anySuperHeroId: string = "Gambito";

export const anyListOfSuperHeroes: SuperHero[] = [
    new SuperHero("Scarlet Witch",
        "https://i.annihil.us/u/prod/marvel/i/mg/9/b0/537bc2375dfb9.jpg",
        false,
        "Scarlet Witch was born at the Wundagore base of the High Evolutionary, she and her twin "
        + "brother Pietro were the children of Romani couple Django and Marya Maximoff. The "
        + "High Evolutionary supposedly abducted the twins when they were babies and "
        + "experimented on them, once he was disgusted with the results, he returned them to"
        + " Wundagore, disguised as regular mutants."),
    new SuperHero("Iron Man",
        "https://i.annihil.us/u/prod/marvel/i/mg/c/60/55b6a28ef24fa.jpg",
        true,
        "Wounded, captured and forced to build a weapon by his enemies, billionaire "
        + "industrialist Tony Stark instead created an advanced suit of armor to save his "
        + "life and escape captivity. Now with a new outlook on life, Tony uses his money "
        + "and intelligence to make the world a safer, better place as Iron Man."),
    new SuperHero("Wolverine",
        "https://i.annihil.us/u/prod/marvel/i/mg/9/00/537bcb1133fd7.jpg",
        false,
        "Born with super-human senses and the power to heal from almost any wound, " +
        "Wolverine "
        + "was captured by a secret Canadian organization and given an unbreakable "
        + "skeleton and claws. Treated like an animal, it took years for him to control"
        + " himself. Now, he's a premiere member of both the X-Men and the Avengers."),
];

export const anyRegularSuperHero = anyListOfSuperHeroes[0];

export const anyAvenger = anyListOfSuperHeroes[1];
