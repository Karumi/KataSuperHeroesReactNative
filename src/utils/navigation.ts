export interface NavigationLeafRoute<Params> {
    params?: Params;
  }
