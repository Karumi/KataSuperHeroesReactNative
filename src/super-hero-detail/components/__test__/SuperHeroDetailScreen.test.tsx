import { none, some } from "fp-ts/lib/Option";
import * as React from "react";
import * as renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { action } from "typesafe-actions";
import { navigationWithMockedParams } from "../../../test-doubles/navigation";
import { anyRegularSuperHero } from "../../../test-doubles/super-heroes";
import SuperHeroDetailScreen from "../SuperHeroDetailScreen";

const anySuperHeroId = "stark";
const anyFetchSuperHeroByIdAction = (id: string) => action(`fetchSuperHeroById - ${id}`);

jest.mock("../../super-hero-detail-actions", () => {
    return {
        fetchSuperHeroById: (() => anyFetchSuperHeroByIdAction(anySuperHeroId)),
    };
});

jest.mock("../../../base-components/loading/Loading", () => {
    return {
        default: "Loading",
    };
});

jest.mock("../../../base-components/empty-case/EmptyCase", () => {
    return {
        default: "EmptyCase",
    };
});

jest.mock("../../../super-heroes-list/components/SuperHeroCell", () => {
    return {
        defatult: "SuperHeroCell",
    };
});
it("invokes the onMount prop when it renders", () => {
    const store = givenAStoreLoadingASuperHero();
    givenTheComponentIsInitializedForASuperHeroWithId(anySuperHeroId);

    renderer.create(
        <SuperHeroDetailScreen store={store} navigation={navigationWithMockedParams} />,
    );

    const actions = store.getActions();
    expect(actions).toEqual([anyFetchSuperHeroByIdAction(anySuperHeroId)]);
});

it("render a loading component while the super hero is being fetched", () => {
    const store = givenAStoreLoadingASuperHero();
    givenTheComponentIsInitializedForASuperHeroWithId(anySuperHeroId);

    const component = renderer.create(
        <SuperHeroDetailScreen store={store} navigation={navigationWithMockedParams} />,
    ).toJSON();

    expect(component).toMatchSnapshot();
});

it("render an empty case component if the super hero fetched does not exist", () => {
    const store = givenAStoreWithAnNonExistingSuperHero();
    givenTheComponentIsInitializedForASuperHeroWithId(anySuperHeroId);

    const component = renderer.create(
        <SuperHeroDetailScreen store={store} navigation={navigationWithMockedParams} />,
    ).toJSON();

    expect(component).toMatchSnapshot();
});

it("renders the super hero loaded", () => {
    const store = givenAStoreWithASuperHeroLoaded();
    givenTheComponentIsInitializedForASuperHeroWithId(anySuperHeroId);

    const component = renderer.create(
        <SuperHeroDetailScreen store={store} navigation={navigationWithMockedParams} />,
    ).toJSON();

    expect(component).toMatchSnapshot();
});

function givenTheComponentIsInitializedForASuperHeroWithId(id: string) {
    navigationWithMockedParams.getParam.mockReturnValue(id);
}

function givenAStoreLoadingASuperHero() {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);
    return mockStore({
        superHeroDetail: {
            loading: true,
            superHero: none,
        },
    });
}

function givenAStoreWithAnNonExistingSuperHero() {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);
    return mockStore({
        superHeroDetail: {
            loading: false,
            superHero: none,
        },
    });
}

function givenAStoreWithASuperHeroLoaded() {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);
    return mockStore({
        superHeroDetail: {
            loading: false,
            superHero: some(anyRegularSuperHero),
        },
    });
}
