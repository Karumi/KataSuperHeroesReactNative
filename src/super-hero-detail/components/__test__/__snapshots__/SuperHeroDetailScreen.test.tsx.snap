// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render a loading component while the super hero is being fetched 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#22282F",
      "flex": 1,
    }
  }
>
  <Loading />
</View>
`;

exports[`render an empty case component if the super hero fetched does not exist 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#22282F",
      "flex": 1,
    }
  }
>
  <EmptyCase />
</View>
`;

exports[`renders the super hero loaded 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#22282F",
      "flex": 1,
    }
  }
/>
`;
